### PR TITLE
Security review basic auth

### PR DESCRIPTION
**Pull request by cyx**
- As requested by @arirubinstein we should add basic auth for POSTs to prevent DOS attacks
- As suggested by @tt it's much better to have multiple auth tokens in preparation for cred rolls.

### CODE REVIEW

**Comment by arirubinstein** [comment]
:+1:

File: util/config.go [hunk 1]
```diff
@@ -9,6 +9,7 @@ import (
 )
 
 var (
+	Tokens             = flag.String("tokens", os.Getenv("TOKENS"), "Colon separated TOKEN(s) to use for authentication.")
```
Comment (tt) [inline util/config.go]: "Colon separated tokens"? It's can't be a separated token...

File: server/server.go [hunk 2]
```diff
@@ -161,8 +161,8 @@ func app() http.Handler {
 	p := pat.New()
 
 	p.GetFunc("/health", addDefaultHeaders(health))
-	p.PostFunc("/streams", addDefaultHeaders(mkstream))
-	p.PostFunc("/streams/:uuid", addDefaultHeaders(pub))
+	p.PostFunc("/streams", auth(addDefaultHeaders(mkstream)))
+	p.PostFunc("/streams/:uuid", auth(addDefaultHeaders(pub)))
```
Comment (tt) [inline server/server.go]: I don't think we agreed to put this on publishing. The remediation was mostly for creating streams.

File: server/middleware.go [hunk 3]
```diff
@@ -27,6 +27,27 @@ func enforceHTTPS(fn http.HandlerFunc) http.HandlerFunc {
...
 }
 
+func auth(fn http.HandlerFunc) http.HandlerFunc {
+	// Empty TOKENS means we don't authenticate
+	// POST requests.
+	if *util.Tokens == "" {
+		return fn
+	}
+
+	tokens := util.StringSliceUtil(strings.Split(*util.Tokens, ":"))
```
Comment (tt) [inline server/middleware.go]: Should this be in the util package?

**Comment by cyx** [comment]
Closing in favor of heroku/authenticater

**Pull request closed without merging**